src/hostapi/pulseaudio/pa_linux_pulseaudio_internal.h

```diff
@@ -0,0 +1,256 @@
+
+/* prototypes for functions declared in this file */
+
+#define PA_PULSEAUDIO_SET_LAST_HOST_ERROR(errorCode, errorText) \
+    PaUtil_SetLastHostErrorInfo(paInDevelopment, errorCode, errorText)
+
+/* Just chosen by hand from mistake and success method. Nothing really groudbreaking
+   If there is better number with better explantion then I'll be glad to change this
+   @todo change this to something more sophisticated */
+#define PULSEAUDIO_TIME_EVENT_USEC 750
+
+#define PAPULSEAUDIO_MAX_DEVICECOUNT 1024
+#define PAPULSEAUDIO_MAX_DEVICENAME 1024
+
+#define PAPULSEAUDIO_FRAMESPERBUFFERUNSPEC 32
+
+/* Assuming of 2 seconds of 44100 Hz sample rate with FLOAT (4 bytes) and stereo channels (2 channels).
+   You should have pretty good size buffer with this. If output/intput doesn't happens in 2 second we
+   have more trouble than this buffer.
+   @todo change this to something more sophisticated */
+#define PULSEAUDIO_BUFFER_SIZE (96100 * 4 * 2)
+
+typedef struct
+{
+    PaUtilHostApiRepresentation inheritedHostApiRep;
+    PaUtilStreamInterface callbackStreamInterface;
+    PaUtilStreamInterface blockingStreamInterface;
+
+    PaUtilAllocationGroup *allocations;
+
+    PaHostApiIndex hostApiIndex;
+    PaDeviceInfo deviceInfoArray[PAPULSEAUDIO_MAX_DEVICECOUNT];
+    char *pulseaudioDeviceNames[PAPULSEAUDIO_MAX_DEVICECOUNT];
+    pa_sample_spec pulseaudioDefaultSampleSpec;
+
+    /* PulseAudio stuff goes here */
+    pa_threaded_mainloop *mainloop;
+    pa_mainloop_api *mainloopApi;
+    pa_context *context;
+    int deviceCount;
+    pa_time_event *timeEvent;
+}
+PaPulseAudio_HostApiRepresentation;
+
+/* PaPulseAudio_Stream - a stream data structure specifically for this implementation */
+
+typedef struct PaPulseAudio_Stream
+{
+    PaUtilStreamRepresentation streamRepresentation;
+    PaUtilCpuLoadMeasurer cpuLoadMeasurer;
+    PaUtilBufferProcessor bufferProcessor;
+    PaPulseAudio_HostApiRepresentation *hostapi;
+
+    unsigned long framesPerHostCallback;
+    pa_threaded_mainloop *mainloop;
+    pa_context *context;
+    pa_sample_spec outputSampleSpec;
+    pa_sample_spec inputSampleSpec;
+    pa_stream *outputStream;
+    pa_stream *inputStream;
+    pa_buffer_attr outputBufferAttr;
+    pa_buffer_attr inputBufferAttr;
+    int outputUnderflows;
+    PaTime latency;
+    int outputChannelCount;
+    int inputChannelCount;
+
+    size_t maxFramesPerBuffer;
+    size_t maxFramesHostPerBuffer;
+    int outputFrameSize;
+    int inputFrameSize;
+
+    PaDeviceIndex inputDevice;
+    PaDeviceIndex outputDevice;
+
+    char *outputStreamName;
+    char *inputStreamName;
+
+    PaUtilRingBuffer inputRing;
+
+    size_t missedBytes;
+
+    /* Used in communication between threads */
+    volatile sig_atomic_t isActive;  /* Is stream in active state? (Between StartStream and StopStream || !paContinue) */
+    volatile sig_atomic_t isStopped; /* Is stream in active state? (Between StartStream and StopStream || !paContinue) */
```

Comment:
The same comment is used to describe both active and stopped.
They should be different.
https://www.portaudio.com/docs/proposals/010-ClarifyStreamStateMachine.html

Reply:
I've added lot longer comment on this and made another commit for adding more commenting. Also cleaned up some old stuff out which is not needed or is not functioning as expected.